### PR TITLE
[GH-526] add copy card link

### PR DESCRIPTION
**Pull request by majidsajadi**
#### Summary
[x] Add a "copy link" option to the card menu, which should simply copy the current URL
[x] Add a "copy link" option to the card menu of the board view, which copy the card URL
[x] Add a "copy link" option to the card menu of the gallery view, which copy the card URL

#### Ticket Link
Fixes https://github.com/mattermost/focalboard/issues/526

### CODE REVIEW

**Comment by majidsajadi** [comment]
@hahmadia @sbishel I used a temporary `add` icon for the menu icon. please provide a `copy` icon so I can replace it.
also didn't write tests since I'm not sure the implementation would be accepted or not.

**Review by sbishel — CHANGES_REQUESTED** [review]
Hi @majidsajadi thanks for your contribution. A couple of issues. 

The  "Copy link" menu item does not line up with other menu items.
Screen Shot 2021-07-29 at 12 22 30 PM

Also, I think we should probably create a single reusable component for the menu rather than duplicating the code in three places. There may be some things to work through, but should be fairly straight forward. Let me know if you have questions regarding this.

@asaadmahmood Could you provide a `copy link` icon?

**Comment by majidsajadi** [comment]
@sbishel about the alignment. It's because the `add` icon is not the same size as the other icons. I will fix this issue when the copy link icon is provided.

about the duplication, do you mean `Menu.Text` or the `onClick` method? if you mean the `onClick` method please let me know your suggestion about where should I create the method. I thought about the `utils` but since it's a card functionality it didn't feel right. also `mutator` is not a good choice since `copyLink` is not mutating anything.

**Comment by sbishel** [comment]
@majidsajadi - I created a separate ticket for [Creating a Reuseable Menu component](https://github.com/mattermost/focalboard/issues/808). But I'm actually talking about the entire `<Menu>` as a component.  But that can be handled separate from this PR.

Once we get the icon, it should be good to go. Thanks for your contribution.

**Comment by majidsajadi** [comment]
@sbishel i would like to work on that issue too.

**Comment by asaadmahmood** [comment]
@hahmadia Here's the icon that we should use:
`icon-link-variant`

You can find it here:
https://mattermost.github.io/compass-icons/

**Comment by asaadmahmood** [comment]
If you need any css help. let me know.

**Comment by hahmadia** [comment]
@asaadmahmood 

I think you meant to tag Scott Bishel @sbishel

**Review by hahmadia — CHANGES_REQUESTED** [review]
Hi @majidsajadi ,

Could you please fix the broken test before I review it. Thank you
```
[tsl] ERROR in /home/runner/work/focalboard/focalboard/webapp/src/components/gallery/galleryCard.tsx(104,82)
      TS2552: Cannot find name 'cardTree'. Did you mean 'cardRef'?
 ```

**Comment by majidsajadi** [comment]
@asaadmahmood the duplicated and delete icons are plain SVG and not compass icons. so should I go with this compass icon for the copy link or do we have SVG for this copy link icon too?

**Comment by asaadmahmood** [comment]
@majidsajadi Ideally we can use the compass icon for all 3, if you're working on the compass icon, just go ahead and update the icon for all 3 of those, that would be great.

**Comment by majidsajadi** [comment]
@asaadmahmood I can do that. but I can't confirm if my changes would break anything or not (especially styles). for example, there are 13 different components that use `DeleteIcon` so if I change this we may encounter some breaks. also `DuplicateIcon`.

**Comment by majidsajadi** [comment]
> @asaadmahmood I can do that. but I can't confirm if my changes would break anything or not (especially styles). for example, there are 13 different components that use `DeleteIcon` so if I change this we may encounter some breaks. also `DuplicateIcon`.

@asaadmahmood  any update? what should I do? update all components? also `icon-link-variant` is not like  `DeleteIcon and `DuplicateIcon`.

**Comment by asaadmahmood** [comment]
@majidsajadi You can update them and then send me a message here, or on mattermost, I can then go and do a quick review to see if things look good, or fix the css where it doesn't.

File: webapp/src/components/gallery/galleryCard.tsx [hunk 1]
```diff
@@ -93,6 +96,16 @@ const GalleryCard = React.memo((props: Props) => {
                                 mutator.duplicateCard(card.id)
                             }}
                         />
+                        <Menu.Text
+                            icon={<LinkIcon/>}
+                            id='copy'
+                            name={intl.formatMessage({id: 'GalleryCard.copyLink', defaultMessage: 'Copy link'})}
+                            onClick={() => {
+                                const url = Utils.addQueryParamToCurrentURL('c', cardTree.card.id)
```
Comment (sbishel) [inline webapp/src/components/gallery/galleryCard.tsx:104]: When this [PR](https://github.com/mattermost/focalboard/pull/1050) goes in, this line should no longer be necessary.

File: webapp/src/components/kanban/kanbanCard.tsx [hunk 2]
```diff
@@ -85,6 +88,16 @@ const KanbanCard = React.memo((props: Props) => {
                                 )
                             }}
                         />
+                        <Menu.Text
+                            icon={<LinkIcon/>}
+                            id='copy'
+                            name={intl.formatMessage({id: 'KanbanCard.copyLink', defaultMessage: 'Copy link'})}
+                            onClick={() => {
+                                const url = Utils.addQueryParamToCurrentURL('c', card.id)
```
Comment (sbishel) [inline webapp/src/components/kanban/kanbanCard.tsx:96]: Same, with recent change, card.id is part of the URL.

**Review by sbishel — CHANGES_REQUESTED** [review]
Let's wait until this [PR](https://github.com/mattermost/focalboard/pull/1050) gets merged and retest, we should be able to remove setting the cardID parameter.

**Comment by hahmadia** [comment]
Also could you please fix the tests, thank you!

File: webapp/src/components/properties/dateRange/__snapshots__/dateRange.test.tsx.snap [hunk 3]
```diff
@@ -97,7 +97,7 @@ exports[`components/properties/dateRange set via text input, es locale 1`] = `
       type="button"
     >
       <span>
-        15 de junio → 20 de junio
+        June 15 → June 20
```
Comment (sbishel) [inline webapp/src/components/properties/dateRange/__snapshots__/dateRange.test.tsx.snap:100]:
```suggestion
        15 de junio → 20 de junio
```

Comment (sbishel) [reply webapp/src/components/properties/dateRange/__snapshots__/dateRange.test.tsx.snap:100]: Causing Unit test failure.

**Review by sbishel — CHANGES_REQUESTED** [review]

**Review by sbishel — APPROVED** [review]

**Comment by sbishel** [comment]
@hahmadia Would you re-review? It won't let me request a review from you, but also shows you requested changes.

**Review by hahmadia — APPROVED** [review]
Hey @majidsajadi ! I'm really sorry for the long over due review. Been busy around here but I finally got a chance to look at it!

All looks great. There was one minor bug where if I was say in Kanban view and clicked `Copy Link`, it would copy the link to the view instead of the card. Knowing how long this card has been in review for , I decided to not let it delay any further and gave a helping hand with addressing it.

image

Otherwise, thank you for the contribution and sorry for taking so long on reviewing this.

**Comment by hahmadia** [comment]
Since there aren't MAJOR changes, I believe this is an appropriate PR for QA to test after it has been merged.

**Pull request merged**